File: src/lib/gh.ts
import { execFile } from "node:child_process";
import { promisify } from "node:util";
import type { WorkflowRun, RunDetail } from "./types.js";

const execFileAsync = promisify(execFile);

async function gh<T>(args: string[]): Promise<T> {
  const { stdout } = await execFileAsync("gh", args, {
    maxBuffer: 10 * 1024 * 1024,
  });
  return JSON.parse(stdout) as T;
}

const RUN_FIELDS = [
  "databaseId",
  "displayTitle",
  "event",
  "headBranch",
  "name",
  "number",
  "status",
  "conclusion",
  "createdAt",
  "updatedAt",
  "url",
  "workflowName",
].join(",");

export async function fetchRuns(repo: string): Promise<WorkflowRun[]> {
  return gh<WorkflowRun[]>([
    "run",
    "list",
    "--repo",
    repo,
    "--json",
    RUN_FIELDS,
    "--limit",
    "20",
  ]);
}

export async function fetchRunDetail(
  repo: string,
  runId: number,
): Promise<RunDetail> {
  return gh<RunDetail>([
    "run",
    "view",
    String(runId),
    "--repo",
    repo,
    "--json",
    `${RUN_FIELDS},jobs`,
  ]);
}

export async function detectRepo(): Promise<string> {
  const { stdout } = await execFileAsync("git", [
    "remote",
    "get-url",
    "origin",
  ]);
  const url = stdout.trim();
  // Handle SSH: git@github.com:owner/repo.git
  const sshMatch = url.match(/git@github\.com:(.+?)(?:\.git)?$/);
  if (sshMatch) return sshMatch[1]!;
  // Handle HTTPS: https://github.com/owner/repo.git
  const httpsMatch = url.match(/github\.com\/(.+?)(?:\.git)?$/);
  if (httpsMatch) return httpsMatch[1]!;
  throw new Error(`Could not parse repo from remote URL: ${url}`);
}

File: internal/gh/gh.go
package gh

import (
	"encoding/json"
	"fmt"
	"os/exec"
	"strconv"

	"github.com/dzoba/github-actions-watcher/internal/types"
)

const runFields = "databaseId,displayTitle,event,headBranch,name,number,status,conclusion,createdAt,updatedAt,url,workflowName"

// FetchRuns returns the 20 most recent workflow runs for a repo.
func FetchRuns(repo string) ([]types.WorkflowRun, error) {
	out, err := exec.Command("gh", "run", "list",
		"--repo", repo,
		"--json", runFields,
		"--limit", "20",
	).Output()
	if err != nil {
		return nil, fmt.Errorf("gh run list failed: %w", err)
	}
	var runs []types.WorkflowRun
	if err := json.Unmarshal(out, &runs); err != nil {
		return nil, fmt.Errorf("failed to parse runs: %w", err)
	}
	return runs, nil
}

// FetchRunDetail returns a single run with its jobs and steps.
func FetchRunDetail(repo string, runID int) (*types.RunDetail, error) {
	out, err := exec.Command("gh", "run", "view",
		strconv.Itoa(runID),
		"--repo", repo,
		"--json", runFields+",jobs",
	).Output()
	if err != nil {
		return nil, fmt.Errorf("gh run view failed: %w", err)
	}
	var detail types.RunDetail
	if err := json.Unmarshal(out, &detail); err != nil {
		return nil, fmt.Errorf("failed to parse run detail: %w", err)
	}
	return &detail, nil
}
